### PR TITLE
Case 21219: Limit new connection rate to the avatar mixer

### PR DESCRIPTION
**Pull request by Atlante45**
[Manuscript Ticket](https://highfidelity.manuscript.com/f/cases/21219/Limit-connection-rate-for-NodeList-new-connections)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17718/HighFidelity-Beta-PR15051-8032063.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17718/HighFidelity-Quest-Beta-PR15051-8032063.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17718/HighFidelity-Beta-PR15051-8032063.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17718/HighFidelity-Beta-Interface-PR15051-8032063.dmg

**Comment by howard-stearns** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17728/HighFidelity-Beta-PR15051-8032063.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17728/HighFidelity-Quest-Beta-PR15051-8032063.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17728/HighFidelity-Beta-PR15051-8032063.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17728/HighFidelity-Beta-Interface-PR15051-8032063.dmg

File: libraries/networking/src/LimitedNodeList.h [hunk 1]
```diff
@@ -440,6 +464,10 @@ private slots:
     Node::LocalID _sessionLocalID { 0 };
     bool _flagTimeForConnectionStep { false }; // only keep track in interface
 
+    size_t _maxConnectionRate { DEFAULT_MAX_CONNECTION_RATE };
+    size_t _nodesAddedInCurrentTimeSlice { 0 };
+    std::vector<NewNodeInfo> _delayedNodeAdds;
```
Comment (SimonWalton-HiFi) [inline libraries/networking/src/LimitedNodeList.h]: I wonder if `_delayedNodeAdds` should be a map or hash keyed on uuid? It's potentially quite large on startup and `std::find_if` etc are just linear searches.

File: libraries/networking/src/LimitedNodeList.h [hunk 2]
```diff
@@ -316,6 +318,9 @@ class LimitedNodeList : public QObject, public Dependency {
     void sendFakedHandshakeRequestToNode(SharedNodePointer node);
 #endif
 
+    size_t getMaxConnectionRate() { return _maxConnectionRate; }
```
Comment (SimonWalton-HiFi) [inline libraries/networking/src/LimitedNodeList.h]: Function should be const.

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17738/HighFidelity-Beta-PR15051-24de8e2.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17738/HighFidelity-Quest-Beta-PR15051-24de8e2.apk)

**Comment by SimonWalton-HiFi** [comment]
Will the default limit (of Agent connection rate) apply to all ACs, with just the avatar mixer having a user option?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17738/HighFidelity-Beta-PR15051-24de8e2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17738/HighFidelity-Beta-Interface-PR15051-24de8e2.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17751/HighFidelity-Beta-PR15051-9fe7a33.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17751/HighFidelity-Quest-Beta-PR15051-9fe7a33.apk)

File: libraries/networking/src/LimitedNodeList.h (hunk 1 above)
Comment (Atlante45) [reply libraries/networking/src/LimitedNodeList.h]: I thought about it, but I want to keep the order of connection and I would lose that with a hash/map.
The vector also stays relatively small 100s of elements so probably not that much more expensive than an STL map.
We also don't check it very often.

**Comment by Atlante45** [comment]
@simon-walton it does apply to all, we could add a disable value or just initialize it to INT_MAX.

**Comment by SimonWalton-HiFi** [comment]
My preference would be to initialize to INT_MAX.

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17756/HighFidelity-Beta-PR15051-a804d35.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17756/HighFidelity-Quest-Beta-PR15051-a804d35.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17756/HighFidelity-Beta-PR15051-a804d35.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17756/HighFidelity-Beta-Interface-PR15051-a804d35.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17756/HighFidelity-Beta-PR15051-a804d35.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17756/HighFidelity-Beta-Interface-PR15051-a804d35.exe

**Pull request merged**